### PR TITLE
Fix bug where records/unions wrapping the same map were not updated when setting a field via the other record/union

### PR DESCRIPTION
**Pull request by karthikrg**

### CODE REVIEW

File: generator-test/src/test/java/com/linkedin/pegasus/generator/override/TestRecord.java [hunk 1]
```diff
@@ -530,4 +530,15 @@ public void testCopyChangeRecordField() throws CloneNotSupportedException
...
+
+  @Test
+  public void testSetOnRecordWrappingSameMap()
+  {
+    RecordBar bar = new RecordBar();
+    bar.setLocation("some");
+    RecordBar copy = new RecordBar(bar.data());
+    assertEquals(copy.getLocation(), "some");
+    copy.setLocation("other");
+    assertEquals(bar.getLocation(), "other");
```
Comment (BrianPin) [inline generator-test/src/test/java/com/linkedin/pegasus/generator/override/TestRecord.java:542]: What is `bar`'s location right now?

File: generator-test/src/test/java/com/linkedin/pegasus/generator/override/TestRecord.java [hunk 2]
```diff
@@ -530,4 +530,15 @@ public void testCopyChangeRecordField() throws CloneNotSupportedException
     assertEquals(record.getRecordOptionalField().getLocation(), "near");
     assertNull(recordCopy.getRecordOptionalField());
   }
+
+  @Test
+  public void testSetOnRecordWrappingSameMap()
+  {
+    RecordBar bar = new RecordBar();
+    bar.setLocation("some");
+    RecordBar copy = new RecordBar(bar.data());
```
Comment (BrianPin) [inline generator-test/src/test/java/com/linkedin/pegasus/generator/override/TestRecord.java:539]: What about deep copy (if we have any)

Comment (karthikrg) [reply generator-test/src/test/java/com/linkedin/pegasus/generator/override/TestRecord.java:539]: Those are already tested via other test cases.

**Review by BrianPin — APPROVED** [review]
LGTM, please think about wc-test on deco and container.
Approve after asking some questions.
The test seems fine to me, please take few minutes to think more if there is corner cases.

**Review by junchuanwang — APPROVED** [review]
